### DISCUSSION TITLE
Lab 02.03 EXTRACT DATA

### DISCUSSION THREAD

**Discussion by jfhr990120** [rramosp/ai4eng.v1#296] [02 - PYTHON]
Hola, espero estén bien.
Resolviendo el primer problema del Lab 02.03 tuve varias dificultades.

Esta es la función que yo creé para obtener la lista de índices que cumplen con las dos condiciones (precio>100 o margen>10)

```
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df2 = df.copy()

    ... # YOUR CODE HERE
    result = df2[(df2.price>100)|(df2.margin>10)]
    result=list(result.index)

    return result
```

Inicialmente encontré el siguiente error
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-70-7a247082c5cc>](https://localhost:8080/#) in <cell line: 1>()
----> 1 select_items(d)

1 frames
[/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py](https://localhost:8080/#) in __getattr__(self, name)
   5900         ):
   5901             return self[name]
-> 5902         return object.__getattribute__(self, name)
   5903 
   5904     def __setattr__(self, name: str, value) -> None:

AttributeError: 'DataFrame' object has no attribute 'margin'
```
Como pueden ver el error está en la creación del dataframe porque no tiene la columna de margin, entonces revisé el código de creacción de la función y cambié una línea de código
Este es el original
```
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)
    
    if missing:
        nmissing = np.random.randint(len(price)//2)+2                                     
        price[np.random.permutation(len(price))[:nmissing]] = np.nan
    
    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]
        
    return d
```
Y yo modifiqué una única línea de código

```
        d = d[d.columns[:3]]
```
Ahora sí me corría el ejemplo y me arrojaba los valores correctos

Pero a la hora de correr el bloque de calificación, el error vuelve a aparecer:

```
error: ERROR EXECUTING YOUR CODE
---------------------------------
Traceback (most recent call last):
  File "/content/calling_script.py", line 24, in <module>
    r = eval("xqwhewcu.{}(source_functions, source_variables, '[jufherreraru@unal.edu.co](mailto:jufherreraru@unal.edu.co)')".format(grader_function_name), namespace)
  File "<string>", line 1, in <module>
  File "/content/xqwhewcu.py", line 46, in grader_01
    answer = st_select_items(d)
  File "<string>", line 6, in select_items
  File "/usr/local/lib/python3.10/dist-packages/pandas/core/generic.py", line 5902, in __getattr__
    return object.__getattribute__(self, name)
AttributeError: 'DataFrame' object has no attribute 'margin'
 (set session.debug=True for tracebacks)
```
Nuevamente hay un problema en el dataframe, pero acá no sé qué hacer porque es el generado por la herramienta de calificación. Si yo cambio mi función y le quito la condición del margen, la herramienta de calificación corre y me muestra el dataframe con la columna llamada margin (adjunto pantallazo, obviamente la calificación es 0 porque no elige todas las filas correctas)

![Error](https://github.com/rramosp/ai4eng.v1/assets/162035477/98f6c1a2-b77a-4f14-bc4a-d178809fccea)

No sé dónde puede estar el problema. Revisé este hilo https://github.com/rramosp/ai4eng.v1/discussions/74 y allí ví un código similar al mío, pero al implementarlo no corre, arrojando el siguiente error

![image](https://github.com/rramosp/ai4eng.v1/assets/162035477/8d1f9d06-b81a-4fe4-948c-282bb2d77f0b)

Agradecería ayuda

@Felipe-RA 
@rramosp

**Comment by jfhr990120** [comment]
Perdón, ya encontré mi error

La columna puede o no puede estar

Ya lo corregí

Disculpen

**Comment by Felipe-RA** [comment]
Vale, me alegra que lo hayas podido solucionar tan rápido!
